google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -442,22 +591,163 @@
    * @param fieldPath The first field to set.
    * @param value The first value to set.
    * @param moreFieldsAndValues String and Object pairs with more fields to be set.
-   * @return An ApiFuture containing the result of the write. Contains an error if the write fails.
+   * @return An ApiFuture containing the result of the write. Contains a {@link BulkWriterError} if
+   *     the write fails.
    */
   @Nonnull
   public ApiFuture<WriteResult> update(
-      @Nonnull DocumentReference documentReference,
-      @Nonnull Precondition precondition,
-      @Nonnull FieldPath fieldPath,
-      @Nullable Object value,
-      Object... moreFieldsAndValues) {
+      @Nonnull final DocumentReference documentReference,
+      @Nonnull final Precondition precondition,
+      @Nonnull final FieldPath fieldPath,
+      @Nullable final Object value,
+      final Object... moreFieldsAndValues) {
     verifyNotClosed();
-    BulkCommitBatch bulkCommitBatch = getEligibleBatch();
-    ApiFuture<WriteResult> future =
-        bulkCommitBatch.update(
-            documentReference, precondition, fieldPath, value, moreFieldsAndValues);
-    sendReadyBatches();
-    return future;
+    return executeWrite(
+        documentReference,
+        OperationType.UPDATE,
+        new BulkWriterOperationCallback() {
+          public ApiFuture<WriteResult> apply(BulkCommitBatch batch) {
+            return batch.update(
+                documentReference, precondition, fieldPath, value, moreFieldsAndValues);
+          }
+        });
+  }
+
+  /**
+   * Schedules the provided write operation and runs the user success callback when the write result
+   * is obtained.
+   */
+  private ApiFuture<WriteResult> executeWrite(
+      final DocumentReference documentReference,
+      final OperationType operationType,
+      final BulkWriterOperationCallback operationCallback) {
+    final SettableApiFuture<Void> operationCompletedFuture = SettableApiFuture.create();
+
+    ApiFuture<WriteResult> writeResultApiFuture =
+        ApiFutures.transformAsync(
+            enqueueWrite(documentReference, operationType, operationCallback),
+            new ApiAsyncFunction<WriteResult, WriteResult>() {
+              public ApiFuture<WriteResult> apply(final WriteResult result)
+                  throws ExecutionException, InterruptedException {
+
+                invokeUserSuccessCallback(documentReference, result).get();
+                return ApiFutures.immediateFuture(result);
+              }
+            },
+            MoreExecutors.directExecutor());
+
+    writeResultApiFuture.addListener(
+        new Runnable() {
+          public void run() {
+            pendingOperations.remove(operationCompletedFuture);
+            operationCompletedFuture.set(null);
+          }
+        },
+        MoreExecutors.directExecutor());
```

Comment:
I think this is where you want to use your own Executor to remove the need to synchronization on the internal data structures.  To make this easier to parse, I would also suggest doing the same in line 637 (just above)